components/script/fetch.rs

```diff
@@ -0,0 +1,165 @@
+    NetTraitsRequestInit {
+        method: request.method.borrow().clone(),
+        url: request.url(),
+        headers: request.headers.borrow().clone(),
+        unsafe_request: request.unsafe_request,
+        same_origin_data: request.same_origin_data.get(),
+        body: request.body.borrow().clone(),
+        destination: request.destination,
+        synchronous: request.synchronous,
+        mode: request.mode,
+        use_cors_preflight: request.use_cors_preflight,
+        credentials_mode: request.credentials_mode,
+        use_url_credentials: request.use_url_credentials,
+        // TODO: NetTraitsRequestInit and NetTraitsRequest have different "origin"
+        // ... NetTraitsRequestInit.origin: Url
+        // ... NetTraitsRequest.origin: RefCell<Origin>
+        origin: request.url(),
+        referrer_url: from_referer_to_referer_url(&request),
+        referrer_policy: request.referrer_policy.get(),
+        pipeline_id: request.pipeline_id.get(),
+    }
+}
+
+// https://fetch.spec.whatwg.org/#fetch-method
+#[allow(unrooted_must_root)]
+pub fn Fetch(global: GlobalRef, input: RequestOrUSVString, init: &RequestInit) -> Fallible<Rc<Promise>> {
+    let core_resource_thread = global.core_resource_thread();
+
+    // Step 1
+    let promise = Promise::new(global);
+    let response = Response::new(global);
+
+    // Step 3
+    response.Headers().set_guard_immutable();
+
+    // Step 2
+    let request = match Request::Constructor(global, input, init) {
+        Err(e) => {
+            promise.maybe_reject_error(promise.global().r().get_cx(),
+                                       e);
+            return Ok(promise);
+        }
+        Ok(r) => r.get_request(),
+    };
+
+    let request_init = request_init_from_request(request);
+
+    // Step 4
+    let (action_sender, action_receiver) = ipc::channel().unwrap();
+    let fetch_context = Arc::new(Mutex::new(FetchContext {
+        fetch_promise: Some(TrustedPromise::new(promise.clone())),
+        response_object: Trusted::new(response.r()),
+    }));
+    let listener = NetworkListener {
+        context: fetch_context.clone(),
+        // TODO: double check how to get script_chan
+        script_chan: global.networking_task_source(),
+        wrapper: None,
+    };
+
+    ROUTER.add_route(action_receiver.to_opaque(), box move |message| {
+        listener.notify_fetch(message.to().unwrap());
+    });
+    core_resource_thread.send(NetTraitsFetch(request_init, action_sender)).unwrap();
+
+    Ok(promise)
+}
+
+impl PreInvoke for FetchContext {}
+
+impl FetchResponseListener for FetchContext {
+    fn process_request_body(&mut self) {
+        // TODO
+    }
+
+    fn process_request_eof(&mut self) {
+        // TODO
+    }
+
+    #[allow(unrooted_must_root)]
+    fn process_response(&mut self, metadata: Result<Metadata, NetworkError>) {
+        let promise = match self.fetch_promise.take() {
```

Comment:
Let's use `let promise = self.fetch_promise.take().expect("fetch promise is missing").root();`